pkg/reconciler/tenancy/initialization/apibinder_initializer_reconcile.go

```diff
@@ -0,0 +1,273 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package initialization
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/kcp-dev/logicalcluster/v2"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog/v2"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/apis/tenancy/initialization"
+	tenancyv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/tenancy/v1alpha1"
+	conditionsv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/third_party/conditions/apis/conditions/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/apis/third_party/conditions/util/conditions"
+	"github.com/kcp-dev/kcp/pkg/logging"
+)
+
+func (b *APIBinder) reconcile(ctx context.Context, clusterWorkspace *tenancyv1alpha1.ClusterWorkspace) error {
+	logger := klog.FromContext(ctx)
+
+	var errors []error
+	clusterName := logicalcluster.From(clusterWorkspace).Join(clusterWorkspace.Name)
+	logger.V(2).Info("initializing APIBindings for workspace")
+
+	// Start with the ClusterWorkspaceType specified by the ClusterWorkspace
+	leafCWT, err := b.getClusterWorkspaceType(logicalcluster.New(clusterWorkspace.Spec.Type.Path), string(clusterWorkspace.Spec.Type.Name))
+	if err != nil {
+		logger.Error(
+			err,
+			"error getting ClusterWorkspaceType",
+			"clusterWorkspaceType.path", clusterWorkspace.Spec.Type.Path,
+			"clusterWorkspaceType.name", clusterWorkspace.Spec.Type.Name,
+		)
+
+		conditions.MarkFalse(
+			clusterWorkspace,
+			tenancyv1alpha1.WorkspaceInitialized,
+			tenancyv1alpha1.WorkspaceInitializedClusterWorkspaceTypeInvalid,
+			conditionsv1alpha1.ConditionSeverityError,
+			"error getting ClusterWorkspaceType %s|%s: %v",
+			clusterWorkspace.Spec.Type.Path, clusterWorkspace.Spec.Type.Name,
+			err,
+		)
+
+		return nil
+	}
+
+	// Get all the transitive ClusterWorkspaceTypes
+	cwts, err := b.transitiveTypeResolver.Resolve(leafCWT)
+	if err != nil {
+		logger.Error(
+			err,
+			"error resolving transitive types",
+			"clusterWorkspaceType.path", clusterWorkspace.Spec.Type.Path,
+			"clusterWorkspaceType.name", clusterWorkspace.Spec.Type.Name,
+		)
+
+		conditions.MarkFalse(
+			clusterWorkspace,
+			tenancyv1alpha1.WorkspaceInitialized,
+			tenancyv1alpha1.WorkspaceInitializedClusterWorkspaceTypeInvalid,
+			conditionsv1alpha1.ConditionSeverityError,
+			"error resolving transitive set of cluster workspace types: %v",
+			err,
+		)
+
+		return nil
+	}
+
+	// Get current bindings
+	bindings, err := b.listAPIBindings(clusterName)
+	if err != nil {
+		errors = append(errors, err)
+	}
+
+	// This keeps track of any APIBindings for an APIExport that may have conflicts. We don't really expect this to
```

Comment:
@sttts might have opinions - this flow makes sense, but it's a lot of bespoke code, and generally could be implemented with the client-go mutation cache + an indexer in a more "common" flow. Deterministic names for the bindings (as opposed to generate-name) would allow for a simpler impl as well.

Reply:
happy to go the mutation cache route if desired.

we need a deterministic, idempotent way to turn a (path, name) reference into a metadata.name for an APIBinding. should this be:
- `<path.replaceAll(":" -> "-")>-<name>`
    - what if this exceeds the 253 max name length?
- a hash of path+name
- other options

?